api/login.go

```diff
@@ -0,0 +1,440 @@
+// Copyright 2013 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package api
+
+import (
+	"context"
+	"net/url"
+	"os"
+	"runtime/debug"
+
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/bakery"
+	"github.com/go-macaroon-bakery/macaroon-bakery/v3/httpbakery"
+	"github.com/juju/errors"
+	"github.com/juju/featureflag"
+	"github.com/juju/names/v5"
+	"github.com/juju/utils/v3"
+	"github.com/juju/version/v2"
+	"gopkg.in/macaroon.v2"
+
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/feature"
+	"github.com/juju/juju/rpc"
+	"github.com/juju/juju/rpc/params"
+	jujuversion "github.com/juju/juju/version"
+)
+
+var (
+	loginDeviceAPICall = func(caller base.APICaller, request interface{}, response interface{}) error {
+		return caller.APICall("Admin", 4, "", "LoginDevice", request, response)
+	}
+	getDeviceSessionTokenAPICall = func(caller base.APICaller, request interface{}, response interface{}) error {
+		return caller.APICall("Admin", 4, "", "GetDeviceSessionToken", request, response)
+	}
+	loginWithSessionTokenAPICall = func(caller base.APICaller, request interface{}, response interface{}) error {
+		return caller.APICall("Admin", 4, "", "LoginWithSessionToken", request, response)
+	}
+	loginWithClientCredentialsAPICall = func(caller base.APICaller, request interface{}, response interface{}) error {
+		return caller.APICall("Admin", 4, "", "LoginWithClientCredentials", request, response)
+	}
+)
+
+// NewSessionTokenLoginProvider returns a LoginProvider implementation that
+// authenticates the entity with the session token.
+func NewSessionTokenLoginProvider(
+	token string,
+	printOutputFunc func(string, ...any) error,
+	updateAccountDetailsFunc func(string) error,
+) *sessionTokenLoginProvider {
+	return &sessionTokenLoginProvider{
+		sessionToken:             token,
+		printOutputFunc:          printOutputFunc,
+		updateAccountDetailsFunc: updateAccountDetailsFunc,
+	}
+}
+
+type sessionTokenLoginProvider struct {
```

Comment:
Can we move these login provider implementations to a different file `loginproviders.go`, same with tests.
Or better a separate go file per provider to keep things nice and neat.

Reply:
done